File: CSharp/SpecStudioExamplesTestsCSharp/common/OrderItemTyped.cs
namespace SpecStudioExamplesTestsCSharp.common
{
using System;
using System.Buffers;
using System.Collections.Generic;
using System.Globalization;
using System.Text;
using System.Text.Json;
using production;

    public class OrderItemTyped
    {
        public SimpleText name;
        public int quantity;
        public Dollar price;
        public Dollar itemTotal;

        public OrderItemTyped(SimpleText name, int quantity, Dollar price, Dollar itemTotal)
        {
            this.name = name;
            this.quantity = quantity;
            this.price = price;
            this.itemTotal = itemTotal;
        }

        public void WriteJson(Utf8JsonWriter w)
        {
            w.WriteStartObject();
            w.WriteString("name", Json.ToText(this.name));
            w.WriteNumber("quantity", this.quantity);
            w.WriteString("price", Json.ToText(this.price));
            w.WriteString("itemTotal", Json.ToText(this.itemTotal));
            w.WriteEndObject();
        }

        public string ToJSON()
        {
            var buffer = new ArrayBufferWriter<byte>();
            using (var w = new Utf8JsonWriter(buffer)) { WriteJson(w); }
            return Encoding.UTF8.GetString(buffer.WrittenSpan);
        }

        public static OrderItemTyped FromJsonElement(JsonElement m)
        {
            return new OrderItemTyped(
                new SimpleText(Json.AsString(Json.Require(m, "name"), "name")),
                Json.AsInt(Json.Require(m, "quantity"), "quantity"),
                new Dollar(Json.AsString(Json.Require(m, "price"), "price")),
                new Dollar(Json.AsString(Json.Require(m, "itemTotal"), "itemTotal"))
            );
        }

        public static OrderItemTyped FromJSON(string json)
        {
            return FromJsonElement(Json.Parse(json));
        }

        public static string ToJSONList(List<OrderItemTyped> list)
        {
            var buffer = new ArrayBufferWriter<byte>();
            using (var w = new Utf8JsonWriter(buffer))
            {
                w.WriteStartArray();
                foreach (var item in list) item.WriteJson(w);
                w.WriteEndArray();
            }
            return Encoding.UTF8.GetString(buffer.WrittenSpan);
        }

        public static List<OrderItemTyped> FromJSONList(string json)
        {
            var result = new List<OrderItemTyped>();
            var root = Json.Parse(json);
            Json.RequireArray(root, "OrderItemTyped");
            foreach (var e in root.EnumerateArray()) result.Add(FromJsonElement(e));
            return result;
        }
    }
}

File: Java/SpectableJavaExamplesTest/src/test/java/spectable/common/OrderItemTyped.java
package spectable.common;

import java.util.ArrayList;
import java.util.List;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import production.*;
import records.*;
import calculator.*;

public class OrderItemTyped {
    public SimpleText name;
    public int quantity;
    public Dollar price;
    public Dollar itemTotal;

    public OrderItemTyped(SimpleText name, int quantity, Dollar price, Dollar itemTotal) {
        this.name = name;
        this.quantity = quantity;
        this.price = price;
        this.itemTotal = itemTotal;
    }

    public OrderItemTyped(OrderItemString s) {
        this.name = new SimpleText(s.name);
        this.quantity = Integer.parseInt(s.quantity);
        this.price = new Dollar(s.price);
        this.itemTotal = new Dollar(s.itemTotal);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof OrderItemTyped)) return false;
        OrderItemTyped that = (OrderItemTyped) o;
        return Objects.equals(name, that.name)
            && Objects.equals(quantity, that.quantity)
            && Objects.equals(price, that.price)
            && Objects.equals(itemTotal, that.itemTotal);
    }

    @Override
    public int hashCode() {
        return Objects.hash(name, quantity, price, itemTotal);
    }

    public OrderItemString toOrderItemString() {
        return new OrderItemString(String.valueOf(name), String.valueOf(quantity), String.valueOf(price), String.valueOf(itemTotal));
    }

    public static List<OrderItemTyped> fromStringList(List<OrderItemString> list) {
        List<OrderItemTyped> result = new ArrayList<>();
        for (OrderItemString s : list) result.add(new OrderItemTyped(s));
        return result;
    }

    public static List<OrderItemString> toStringList(List<OrderItemTyped> list) {
        List<OrderItemString> result = new ArrayList<>();
        for (OrderItemTyped t : list) result.add(t.toOrderItemString());
        return result;
    }

    /** Internal plumbing for {@link Json}; use toJSON() for JSON text. */
    public Map<String, Object> toJsonValue() {
        Map<String, Object> m = new LinkedHashMap<String, Object>();
        m.put("name", name == null ? null : name.toString());
        m.put("quantity", quantity);
        m.put("price", price == null ? null : price.toString());
        m.put("itemTotal", itemTotal == null ? null : itemTotal.toString());
        return m;
    }

    public String toJSON() { return Json.write(toJsonValue()); }

    /** Internal plumbing for {@link Json}; use fromJSON(String) for JSON text. */
    public static OrderItemTyped fromJsonValue(Map<String, Object> m) {
        if (m == null) return null;
        return new OrderItemTyped(
            new SimpleText(Json.asString(Json.require(m, "name"), "name")),
            Json.asInt(Json.require(m, "quantity"), "quantity"),
            new Dollar(Json.asString(Json.require(m, "price"), "price")),
            new Dollar(Json.asString(Json.require(m, "itemTotal"), "itemTotal")));
    }

    public static OrderItemTyped fromJSON(String json) {
        return fromJsonValue(Json.parseObject(json));
    }

    public static String toJSONList(List<OrderItemTyped> list) {
        List<Object> arr = new ArrayList<Object>();
        if (list != null)
            for (OrderItemTyped item : list) arr.add(item == null ? null : item.toJsonValue());
        return Json.write(arr);
    }

    public static List<OrderItemTyped> fromJSONList(String json) {
        List<OrderItemTyped> result = new ArrayList<OrderItemTyped>();
        for (Object e : Json.parseArray(json))
            result.add(fromJsonValue(Json.asObject(e, "OrderItemTyped")));
        return result;
    }
}
